### PR TITLE
fix(augmentation): let RandomTransplantation run after other steps in a container

### LINKED ISSUE #4707: RandomTransplantation inside AugmentationSequential raises unless it is the first op: an earlier child promotes the mask to (B, 1, H, W)
## Summary

`RandomTransplantation` inside `AugmentationSequential` works **only when no other augmentation runs
before it**. Any earlier `_AugmentationBase` child — even one whose gate is closed — leaves the mask as
`(B, 1, H, W)`, and the transplant then refuses it, because it requires the image to have exactly one
more axis than the mask. The container's own documented mask layout `(B, 1, H, W)` is refused for the
same reason even when the transplant is the first op.

The class docstring says it is "intended to be used with `AugmentationSequential`", and a realistic
pipeline (flip or normalize, then transplant) is exactly the case that raises.

## Repro

```python
import torch
import kornia.augmentation as K

img = torch.rand(3, 2, 4, 6)
mask = torch.randint(0, 3, (3, 4, 6))
T = lambda: K.RandomTransplantation(p=1.0)

def run(name, *ops, m=mask):
    try:
        out = K.AugmentationSequential(*ops, data_keys=["image", "mask"])(img, m)
        print(name, "OK", [tuple(o.shape) for o in out])
    except Exception as e:
        print(name, "RAISED", type(e).__name__, str(e)[:95])

run("transplant alone            ", T())
run("HFlip(p=0) then transplant  ", K.RandomHorizontalFlip(p=0.0), T())
run("Normalize then transplant   ", K.Normalize(0.5, 0.5), T())
run("transplant then HFlip(p=0)  ", T(), K.RandomHorizontalFlip(p=0.0))
run("transplant alone, (B,1,H,W) ", T(), m=mask[:, None])
```

## Observed (main @ 5390ec29d, torch 2.14.0, CPU)

```
transplant alone             OK [(3, 2, 4, 6), (3, 4, 6)]
HFlip(p=0) then transplant   RAISED BaseError Every image input must have one additional dimension (channel dimension) than the segmentation mask, but got 4 for the input image and 4 for the segmentation mask.
Normalize then transplant    RAISED BaseError (same message)
transplant then HFlip(p=0)   OK [(3, 2, 4, 6), (3, 1, 4, 6)]
transplant alone, (B,1,H,W)  RAISED BaseError (same message)
```

The traceback ends in `RandomTransplantation.params_from_input`, called from
`kornia/augmentation/container/ops.py` (`AugmentationSequentialOps.transform`). The earlier child's
mask op has already promoted the mask to `(B, 1, H, W)`, so `_input.ndim == mask.ndim + 1` fails.
A preceding `nn.Identity` does not trigger it. `RandomTransplantation3D` behaves the same way with
`(B, 1, D, H, W)`.

Two smaller consequences of the same mismatch:

- the returned mask rank depends on the pipeline: `(B, H, W)` when the transplant is the only op,
  `(B, 1, H, W)` when anything follows it. The `AugmentationSequential` docstring says a `(B, H, W)`
  mask "is accepted and returned as `(B, 1, H, W)`", which is not what a transplant-only pipeline does.
- the check that fails is a rank check, so the message talks about a missing channel dimension on the
  *image*, which is not what is wrong.

## Not a regression

Identical on the `v0.7.3` and `v0.8.1` tags (same repro, `Exception: False not true. Every image input
must have one additional dimension ...`), so this has been the behaviour since the class was added.
The existing container test for the class, `test_repeating[AugmentationSequential]`, only ever builds
a pipeline with the transplant as its single op.

## Classification

A defect, not a convention: the documented intended use raises. The likely fix is on the
transplantation side — accept a mask carrying a singleton channel axis (`mask.ndim == image.ndim` with
`mask.shape[1] == 1`) and select through it — rather than in the container, since every other mask
consumer already lives with the promoted layout. Found while reviewing #4695 (batch 6e of #4407);
that PR pins the current behaviour as a wart and does not change it. Related container-scope items from
the same audit: #4692, #4693.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by dchaudhari7177**
## What does this pull request do?

This fixes #4707 on the transplantation side, as the issue suggests.

Once any earlier child has run, `AugmentationSequential` hands the mask on as `(B, 1, H, W)`, the same rank as the image. `RandomTransplantation.params_from_input` required `image.ndim == mask.ndim + 1`, so it refused that layout, and the transplant only worked as the first step. Passing a `(B, 1, H, W)` mask directly was refused the same way.

**Now:**

- In `params_from_input`, when the driving mask has a singleton channel axis and an image of the same rank is in the call, the labels and `selection` are built from `mask.squeeze(1)`. The selection is exactly the one a `(B, H, W)` mask gives.
- In `transform_mask`, when a mask is one rank above the selection, the singleton axis is re-added (`unsqueeze(1).expand_as(donor)`). Each mask therefore comes back in its own layout.
- A mask-only call (no image) keeps its own rank, so a `(B, 1, W)` mask is still read as spatial `(1, W)` and isn't squeezed.
- `RandomTransplantation3D` inherits both paths.

**Issue repro, before → after:**

```
transplant alone             OK [(3, 2, 4, 6), (3, 4, 6)]    (unchanged)
HFlip(p=0) then transplant   RAISED  ->  OK [(3, 2, 4, 6), (3, 1, 4, 6)]
Normalize then transplant    RAISED  ->  OK [(3, 2, 4, 6), (3, 1, 4, 6)]
transplant then HFlip(p=0)   OK [(3, 2, 4, 6), (3, 1, 4, 6)] (unchanged)
transplant alone, (B,1,H,W)  RAISED  ->  OK [(3, 2, 4, 6), (3, 1, 4, 6)]
```

**Not changed:** the mask rank a transplant-only pipeline returns, `(B, H, W)` versus `(B, 1, H, W)`, which is the first of the issue's "smaller consequences". Changing that alters output shapes, so I've left it for a separate decision. The Convention block now documents the singleton-channel reading and no longer says the transplant only works as the first step.

Changelog fragment: `changelog.d/4740.fixed.md`.

## Related issue or discussion

Fixes #4707. #4695 pinned the refusal as `test_wart_container_runs_the_transplant_only_as_the_first_step_4707`, and this PR replaces that wart test.

## What did you check?

```text
$ python -m pytest tests/augmentation/test_conventions_transplantation.py tests/augmentation/container \
    tests/augmentation/test_augmentation_mix.py tests/augmentation/test_conventions_mix.py
1094 passed, 235 skipped, 2 xfailed
$ python -m pytest tests/augmentation/test_base.py tests/augmentation/test_conventions_3d.py \
    tests/augmentation/test_conventions_6d_serialization.py
195 passed, 4 xfailed
```

New tests in `test_conventions_transplantation.py`:

- `test_convention_container_runs_the_transplant_after_another_step_4707[closed-gate-flip|normalize]`: runs the pipeline, then replays the container's recorded `_params` in a direct call on the `(B, H, W)` mask (and the normalized image) and checks that the same positions moved. The mask shape is `(B, 1, H, W)` and the selection shape is `(B, H, W)`.
- `test_convention_a_singleton_channel_mask_is_the_spatial_mask_4707[2d|3d]`: a direct call with `(B, 1, *spatial)` under the same seed matches the `(B, *spatial)` result.
- `test_convention_a_mask_only_call_keeps_its_own_rank_4707`: control for the mask-only case.

With `transplantation.py` reverted, the first four fail and the control passes. The doctests pass, and `ruff check` / `ruff format --check` at the pre-commit pin are clean. torch 2.5.0, CPU.

## How did AI help?

I used an AI assistant to trace the container's mask promotion and to draft the tests. I ran the issue's repro before and after, checked the red/green results above, and reviewed the diff myself.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
## Review at `926ebc026356397a889a89ad3c5e7f883706c61e`

The approach is right and matches what #4707 suggested: squeeze the driving mask's singleton channel only when an image of the same rank is in the call, then re-add the axis per mask in `transform_mask`. The container repros from the issue now pass on 2D, and a direct `(B, 1, D, H, W)` call on `RandomTransplantation3D` matches the `(B, D, H, W)` result. Merge-tree is clean against main and against #4697, which also touches `_3d/mix/transplantation.py`. #4697's only `#4707` hit is the Convention sentence this PR rewrites, and no stale `#4707`/`4707` prose or pins are left in `kornia/`, `tests/` or `docs/`. Two items block.

**Verified** (on the branch merged with main `8b03a8085`, run from the worktree, `kornia.__file__` checked):
- `test_conventions_transplantation.py`, `tests/augmentation/container`, `test_augmentation_mix.py`, `test_conventions_mix.py`, `test_conventions_3d.py`, `tests/augmentation/_2d`, `tests/augmentation/_3d` (float32): 1221 passed, 0 failed. Doctests pass.
- A container with two mask keys, `(B, H, W)` and `(B, 1, H, W)`, behind a closed-gate flip moves both identically.
- Mutation testing on `transplantation.py` over the same test set:
  - squeeze replaced by `pass`: 4 failed (killed)
  - `_input.ndim == mask.ndim` changed to `!=`: 5 failed (killed)
  - the `any(...)` image clause dropped: 2 failed (killed)
  - `DataKey.INPUT` changed to `DataKey.MASK` in that clause: 2 failed (killed)
  - `mask.ndim > 1` changed to `>=`: 1 failed (killed)
  - `transform_mask` condition changed to `ndim == selection.ndim`: 80 failed (killed)
  - `mask.shape[1] == 1` changed to `>= 1`: survives, but it is equivalent, because `squeeze(1)` is a no-op on a non-singleton axis and the rank check still refuses it
  - **`.expand_as(donor)` dropped in `transform_mask`: survives, and it is not equivalent (item 2)**

### Blocks

**1. The changelog fragment overclaims for `RandomTransplantation3D`.** It says both classes "no longer raise when another augmentation runs before them" inside `AugmentationSequential`. For 3D that is unreachable. Every other 3D child raises `NotImplementedError` on a mask key, on main and on this branch: `RandomHorizontalFlip3D(p=0)` raises "The support for 3d mask operations are not yet supported" (`container/ops.py`), and `RandomEqualize3D(p=0)` / `RandomMotionBlur3D(p=0)` raise a bare `NotImplementedError`. The 3D change is real, but it only applies to a direct call with a `(B, 1, D, H, W)` mask. Suggested wording:

```
`RandomTransplantation` and `RandomTransplantation3D` accept a `(B, 1, *spatial)` mask next to an image of the same rank and read it as the `(B, *spatial)` mask, so `RandomTransplantation` no longer raises inside `AugmentationSequential` when another augmentation runs before it.
```

**2. Nothing pins the `.expand_as(donor)` in `transform_mask`.** Because of it, this PR also starts accepting a further *multi-channel* mask one rank above the driving mask. For example, `data_keys=["input", "mask", "mask"]` with a `(B, H, W)` driving mask and a `(B, 2, H, W)` second mask moves every channel through the selection. On main that call fails with an indexing error. The Convention bullet only describes the singleton case, and dropping `.expand_as(donor)` leaves the suite green. Either document and pin the per-channel behaviour, or restrict the branch to `acceptor.shape[self._channel_dim] == 1` and drop the expand. If you keep it, this pin passes on the branch, fails with the lib reverted to main, and fails with the expand dropped. It is ruff-clean at the pinned 0.16.8:

```diff
@@ -536,6 +536,19 @@ class TestTransplantationConventions(BaseTester):
         assert out.shape == mask.shape
         assert aug._params["selection"].shape == (3, 1, 5)
 
+    @pytest.mark.device_agnostic
+    def test_convention_a_multi_channel_extra_mask_moves_every_channel_4707(self):
+        # A further mask one rank above the driving mask is read as ``(B, C, *spatial)``: every channel moves
+        # through the same selection as the driving ``(B, *spatial)`` mask.
+        image, mask = _multi_label_batch(batch=3)
+        extra = torch.stack([mask, mask + 10], dim=1)
+        torch.manual_seed(5)
+        _, out_mask, out_extra = K.RandomTransplantation(p=1.0)(image, mask, extra, data_keys=["input", "mask", "mask"])
+        assert out_extra.shape == extra.shape
+        assert not torch.equal(out_mask, mask)  # something really moved
+        assert torch.equal(out_extra[:, 0], out_mask)
+        assert torch.equal(out_extra[:, 1], out_mask + 10)
+
     @pytest.mark.device_agnostic
     @pytest.mark.parametrize(
         "cls, spatial", [(K.RandomTransplantation, (4, 6)), (K.RandomTransplantation3D, (3, 4, 6))], ids=["2d", "3d"]
```

Also add a clause to the new Convention bullet, e.g. "a further mask one rank above the driving mask is read as `(B, C, *spatial)` and every channel moves".

### Filed (not blocking)

- #4750 tracks the two "smaller consequences" of #4707 that this PR leaves on purpose: a transplant-only pipeline returns `(B, H, W)`, which contradicts the `AugmentationSequential` docstring, and the refusal message blames the image. With those tracked, `Fixes #4707` is the right keyword for this PR.

Tests follow the file's existing `device_agnostic` pattern. The transplant is index logic, so no dtype or device fixture is needed.

CI: all checks green.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by dchaudhari7177** [comment]
Thanks. Both are addressed in `f1244db4`.

1. **Fragment.** I used your wording. The container claim now covers only `RandomTransplantation`, and the 3D change is described as the direct `(B, 1, D, H, W)` call.
2. **`.expand_as(donor)`.** I kept the per-channel behaviour. It's the natural reading of a `(B, C, *spatial)` mask, and the driving mask's `selection` is already per pixel. I added your pin `test_convention_a_multi_channel_extra_mask_moves_every_channel_4707` and a Convention clause: "A further mask one rank above the driving mask is read as `(B, C, *spatial)`, and every channel moves through the same `selection`."

Checked locally (torch 2.5.0, CPU, float32):
- Dropping the `.expand_as(donor)` at `transplantation.py:251` fails the new pin. The other 61 in the file pass.
- `test_conventions_transplantation.py`, `tests/augmentation/container` and `test_augmentation_mix.py`: 1029 passed. The doctest passes.
- ruff 0.16.8 check and format are clean.

**Review by ducha-aiki — APPROVED** [review]
## Re-review at `f1244db4f`

Thanks, both blocking items from the last round are fixed. Approving.

**Earlier blocks**
1. **Fragment overclaim: fixed.** `changelog.d/4740.fixed.md` now uses the suggested wording. The container claim covers only `RandomTransplantation`, and for `RandomTransplantation3D` the fragment describes only the direct `(B, 1, D, H, W)` call.
2. **`.expand_as(donor)` pin: fixed.** `test_convention_a_multi_channel_extra_mask_moves_every_channel_4707` is in, and the Convention bullet now says a further mask one rank above the driving mask is read as `(B, C, *spatial)`. If `.expand_as(donor)` is dropped, that pin is the one test that fails. With `transplantation.py` reverted to main, it fails along with the four other `#4707` pins, and the mask-only control still passes.

**Checked again at this head.** I re-read the whole diff. The library code has not changed since `926ebc026`; only the docstring moved.
- The new Convention sentence holds when run. A `(B, 3, *spatial)` extra mask moves every channel through the driving selection for both classes. This works whether the driving mask is `(B, *spatial)` or `(B, 1, *spatial)`, and also in a mask-only call.
- The fix also works in a container after `RandomAffine`, `ColorJitter` and `RandomHorizontalFlip(p=1)` with an asymmetric 5x7 image. It also handles a `bool` `(B, 1, H, W)` mask, `B = 1` (still an identity), an empty batch and a `C = 1` image.
- Mutation testing on `transplantation.py`, run against `test_conventions_transplantation.py`, `test_augmentation_mix.py`, `test_conventions_mix.py` and `tests/augmentation/container`, killed every non-equivalent mutant, including the dropped expand. One mutant survives: `==` changed to `>=` in the `transform_mask` rank check. It only matters for a mask two ranks above the driving mask, which is undocumented and already raises at this head, so it is not a gap.
- `#4707` appears only in the new Convention bullet, the code comments and the regression tests, and no wart prose is left. `#4750` does not appear anywhere in the diff, and the PR does not claim to fix it.
- Merge-tree is clean against current main (`f9129bcea`, which includes #4791), #4739 and #4764. On main + this PR, I ran the transplantation tests, `test_augmentation_mix.py`, `test_conventions_mix.py` and `tests/augmentation/container` at float32 and float64: 2002 passed, 0 failed. On main + this PR + #4764: 875 passed. On main + this PR + #4739: 872 passed.
- On torch 2.5.1, `test_conventions_transplantation.py` at all dtypes gives 65 passed. The doctest passes. Ruff 0.16.8 `check` and `format --check` are clean.

CI: 41/41 checks green.


Posted on behalf of @ducha-aiki by Claude (Opus 5.5).
